frontend/src/metabase/parameters/components/ParameterValuePicker/ListPickerConnected/ListPickerConnected.tsx

```diff
@@ -0,0 +1,107 @@
+import { useCallback, useEffect, useState } from "react";
+import { useAsync } from "react-use";
+import { t } from "ttag";
+import { useDebouncedCallback } from "use-debounce";
+
+import type { Parameter, ParameterValues } from "metabase-types/api";
+
+import { ListPicker } from "../ListPicker";
+import {
+  getFlattenedStrings,
+  getListParameterStaticValues,
+  shouldEnableSearch,
+} from "../core";
+
+interface ListPickerConnectedProps {
+  value: string | null;
+  parameter: Parameter;
+  onChange: (value: string | null) => void;
+  fetchValues: (query: string) => Promise<ParameterValues>;
+  forceSearchItemCount: number;
+  searchDebounceMs?: number;
+}
+
+export function ListPickerConnected(props: ListPickerConnectedProps) {
+  const {
+    value,
+    parameter,
+    onChange,
+    forceSearchItemCount,
+    searchDebounceMs = 150,
+    fetchValues,
+  } = props;
+
+  const [isOpen, setIsOpen] = useState(false);
+  const [_, setSearchValue] = useState("");
+  const [searchQuery, setSearchQuery] = useState("");
+
+  useEffect(() => {
+    onChange(null);
+    setIsOpen(false);
+    setSearchQuery("");
+    setSearchValue("");
+  }, [onChange, parameter.values_source_type, parameter.values_source_config]);
+
+  const {
+    value: fetchedValues,
+    loading,
+    error,
+  } = useAsync(async () => {
+    if (isOpen) {
+      return await fetchValues(searchQuery);
+    }
+    return undefined;
+  }, [
+    isOpen,
+    searchQuery,
+    parameter.values_source_type,
+    parameter.values_source_config,
+    fetchValues,
+  ]);
```

Comment:
@ranquild I have a few questions I don't know how to solve elegantly with this approach.

1. We need to prevent re-fetches when opening and closing the dropdown. Opening it re-renders the component because of `setIsOpen` but having `isOpen` as a hook dependency is needed to prevent fetching when dropdown is closed.

2. When `has_more_values` in the response is `false`, we don't want to attempt fetching data anymore _unless_ the parameter type/value source changes (this part should be taken care of by the "reset" effect). This means we will have to save the previously fetched data and return it when the flag is on unless you have a better idea? 

3. When a value is selected, we need to prevent re-fetching the data. Imagine having `[A, B, C, D]` in a dropdown and selecting `A`. This triggers `onChange` and `onSearchChange` of the `Select` in `ListPicker`, which means you end up seeing `[A]` _and_ unnecessarily fetching it the 2nd time.

I believe there might be some other subtle bugs, which I can't articulate really well right now, but these are the most important ones.

Reply:
2 - should be possible to fix in what is called `useDebouncedCallback` here. You could skip setting the search query if it's known you don't need to search again.

3 - is it because the dropdown is closed when selecting a value? if that's the case could setting `isOpen` flag help?

1 - is `fetchValues` call with exactly the same values cached? It seems to be based on `parameters/actions.ts` but I haven't checked. If it is, we could rely on this.  If it's not cached or we don't want to rely on this, we could do something different

Reply:
1. Yes, these are calls with the same arguments, so we could rely on them. Done.
2. Done.
3. Seemingly I have solved it by introducing another value, `lastSearchChange`.

Reply:
I believe it should be done but please let me know @ranquild if you see any further improvements.